Support clearing devices keys of exported TTNv2 devices

neoaggelos opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

References https://github.com/TheThingsIndustries/lorawan-stack/issues/2417

#### Changes
<!-- What are the changes made in this pull request? -->

- Add `--ttnv2.clear-session` flag.
- Update README with a notice to the users.

#### Testing

<!-- How did you verify that this change works? -->

Test with `--ttnv2.clear-session` flag: Device session is deleted from TTN V2 after exporting.

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

Flag is disabled by default. Users must pass the flag explicitly, in order to avoid clearing the session keys by mistake.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.

johanstokking reviewed (commented):
I would clear the session by default, to be honest. It is really problematic if the session keys are in both servers.

I would have a flag `--dry-run` to indicate that it shouldn't clear, so people can see what's going on exactly.

---

Also, what happens on join?

neoaggelos commented:
Added a `--dry-run` flag, renamed `--ttnv2.clear-session` to `--ttnv2.clear-device-keys` and changed to clear the AppKey as well. This flag is also enabled by default, and the docs explicitly mention to verify any operations with `--dry-run` before actually running them.

johanstokking commented on README.md:
```diff
@@ -54,12 +54,22 @@ Private The Things Network Stack V2 deployments are also supported, and require
+- The export process will halt if any errors occur. Execute any commands with the `--dry-run` flag first, to verify their result.
+- Payload formatters are not exported. See [Payload Formatters](https://thethingsstack.io/integrations/payload-formatters/).
+- Active device sessions are exported by default. You can disable this by using the `--ttnv2.with-session=false` flag. It is recommended that you do not export session keys for devices that can instead re-join on The Things Stack.
+- **IMPORTANT**: The migration from The Things Network Stack V2 to The Things Stack is one-way. Note that it is crucial that devices are handled by one Network Server at a time. In the commands below, we use the `--ttnv2.clear-device-keys=true` flag, which will clear both the root keys (**AppKey**, if any) the session keys (**AppSKey**, **NwkSKey**) from The Things Network Stack V2 after exportin ...
```
Okay so `clear-device-keys` is by default `false`, but `dry-run` is by default also `false`?

Why don't we have one flag, `dry-run`, and if it's set, don't clear anything? And let `with-session` indicate what to clear and export.

So basically

`--dry-run` | `--with-session` | What happens
--- | --- | ---
`false` | `false` | reset root key, export root key but not session
`false` | `true` | reset session keys and root key, export all
`true` | `false` | no reset, export root key but not session
`true` | `true` | no reset, export all

johanstokking commented on README.md:
````diff
@@ -75,22 +85,23 @@ device5
 ```bash
-$ ttn-lw-migrate device --source ttnv2 < device_ids.txt > devices.json
+# dry run first, verify that no errors occur
+$ ttn-lw-migrate devices --source ttnv2 --ttnv2.clear-device-keys=true "mydevice" --dry-run --verbose < device_ids.txt > devices.json
````
This is where I find `clear-device-keys` confusing with `dry-run`.

johanstokking commented on cmd/root.go:
```diff
@@ -59,4 +59,5 @@ func Execute() int {
 
 func init() {
 	rootCmd.PersistentFlags().Bool("verbose", false, "Verbose output")
+	rootCmd.PersistentFlags().Bool("dry-run", false, "Dry run")
```
Indicate what this means

rvolosatovs replied:
For reference, from `man 1 git push`:
```
       -n, --dry-run
           Do everything except actually send the updates.

```

htdvisser reviewed (commented):
What exactly is the use case for clearing all device fields instead of just deleting the device from v2?

As far as I can think of, there are 3 scenarios:

- I want a dry run, where the CLI tells me what it will do.
- I want to migrate a device, but keep the session on v2 until my device does a new join on v3. In this case we should delete the root keys, but keep the session after migration.
- I want to migrate a device with its session. In this case we can delete the entire device registration from v2 after migration.

johanstokking commented on README.md:
```diff
@@ -54,11 +54,21 @@ Private The Things Network Stack V2 deployments are also supported, and require
 
+### Notes
+
+- The export process will halt if any errors occur. Execute any commands with the `--dry-run` flag first, to verify their result.
```
```suggestion
- The export process will halt if any error occurs.
- Execute commands with the `--dry-run` flag to verify if the outcome will be as expected.
```
